src/nvim/os/input.c

```diff
@@ -323,13 +325,26 @@ static unsigned int handle_mouse_event(char **ptr, uint8_t *buf,
   return bufsize;
 }
 
+bool input_blocking(void)
+{
+  return blocking;
+}
+
 static bool input_poll(int ms)
 {
   if (do_profiling == PROF_YES && ms) {
     prof_inchar_enter();
   }
 
+  if ((ms == - 1 || ms > 100)
+      // && multiqueue_size(main_loop.events) > 1
+      && !(events_enabled || input_ready() || input_eof)
+      ) {
+    blocking = true;
+    multiqueue_process_events(main_loop.events);
```

Comment:
We found out that the pending input will provoke a blocking wait, so here we process queued events (which may include the synchronous `nvim_get_mode`) once before blocking.